File: hello.rs
#![feature(libc)]

extern crate libc;

extern
{
    //fn plus(x: libc::c_int, y: libc::c_int) -> libc::c_int;
    fn plusone(x: libc::c_int) -> libc::c_int;
    fn current_timestamp() -> libc::c_long;
}

use std::env;

fn run(count: libc::c_int)
{
    unsafe
    {
        // start immediately
        let start = current_timestamp();
        
        let mut x = 0;
        while x < count {
            x = plusone(x);
        }
        
        let end = current_timestamp();
        let elapsed = end - start;
        
        println!("{}", elapsed);
    }
}

fn main()
{
    let args: Vec<String> = env::args().collect();
    if args.len() == 1 {
        println!("First arg (0 - 2000000000) is required.");
        return;
    }
    let count = args[1].parse::<i32>().unwrap();
    if count <= 0 || count > 2000000000 {
        println!("Must be a positive number not exceeding 2 billion.");
        return;
    }
    run(count as libc::c_int);
}

File: hello.lua
local ffi = require("ffi")

ffi.cdef[[
int plus(int x, int y);
int plusone(int x);
long long current_timestamp();

int printf(const char *fmt, ...);
]]

-- load
local newplus = ffi.load("./newplus/libnewplus.so")
newplus.current_timestamp()
newplus.plusone(0)

local function run(count)
    local start = newplus.current_timestamp()

    local x = 0
    while x < count do
        x = newplus.plusone(x)
    end
    
    ffi.C.printf("%lld\n", newplus.current_timestamp() - start)
end

local function start()
    if #arg == 0 then
        print("First arg (0 - 2000000000) is required.")
        return
    end

    local count = tonumber(arg[1])
    if count <= 0 or count > 2000000000 then
        print("Must be a positive number not exceeding 2 billion.")
        return
    end
    
    run(count)
end

start()

